src/Products/PluginIndexes/CompositeIndex/CompositeIndex.py

```diff
@@ -0,0 +1,599 @@
+
+        if keywords is not None:
+            for kw in keywords:
+                self.removeForwardIndexEntry(kw, documentId)
+
+    def unindex_object(self, documentId):
+        """ carefully unindex the object with integer id 'documentId'"""
+
+        keywords = self._unindex.get(documentId, None)
+        self.unindex_objectKeywords(documentId, keywords)
+        try:
+            del self._unindex[documentId]
+        except KeyError:
+            LOG.debug('%s: Attempt to unindex nonexistent '
+                      'document with id %s' %
+                      (self.__class__.__name__, documentId),
+                      error=sys.exc_info())
+
+    def _get_object_keywords(self, obj):
+        """ returns permutation list of object keywords """
+
+        components = self.getIndexComponents()
+        kw_list = []
+
+        for c in components:
+            kw = self._get_component_keywords(obj, c)
+            # skip if keyword list is empty
+            if not kw:
+                continue
+            kw = tuple([(c.id, k) for k in kw])
+            kw_list.append(kw)
+
+        # permute keyword list in order to support any combination and
+        # number (n > 1) of components in query
+        pkl = []
+        c_list = product(*kw_list)
+
+        for c in c_list:
+            for r in range(MIN_COMPONENTS, len(c) + 1):
+                p = combinations(c, r)
+                pkl.extend(p)
+
+        return tuple(pkl)
+
+    def _get_component_keywords(self, obj, component):
+
+        if component.meta_type == 'FieldIndex':
+            attr = component.attributes[-1]
+            try:
+                datum = getattr(obj, attr)
+                if safe_callable(datum):
+                    datum = datum()
+            except (AttributeError, TypeError):
+                datum = _marker
+            if isinstance(datum, list):
+                datum = tuple(datum)
+            return (datum,)
+
+        elif component.meta_type == 'KeywordIndex':
+            for attr in component.attributes:
+                datum = []
+                newKeywords = getattr(obj, attr, ())
+                if safe_callable(newKeywords):
+                    try:
+                        newKeywords = newKeywords()
+                    except AttributeError:
+                        continue
+                if not newKeywords and newKeywords is not False:
+                    continue
+                # Python 2.1 compat isinstance
+                elif isinstance(newKeywords, basestring):
+                    datum.append(newKeywords)
+                else:
+                    try:
+                        # unique
+                        newKeywords = set(newKeywords)
+                    except TypeError:
+                        # Not a sequence
+                        datum.append(newKeywords)
+                    else:
+                        datum.extend(newKeywords)
+
+            datum.sort()
+            return tuple(datum)
```

Comment:
Assuming you'd inherit from KeywordIndex, this could be shortened as well:
```
elif component.meta_type == 'KeywordIndex':
    for attr in component.attributes:
        datum = []
        attr_datum = KeywordIndex._get_object_keywords(self, obj, attr)
        if attr_datum:
            if isinstance(attr, tuple):
                datum.append(attr_datum)
            else:
                datum.extend(attr_datum)
        datum.sort()
        return tuple(datum)
```